server/src/main/java/org/opensearch/monitor/fs/FsHealthService.java

```diff
@@ -134,6 +146,9 @@ public StatusInfo getHealth() {
             statusInfo = new StatusInfo(HEALTHY, "health check disabled");
         } else if (brokenLock) {
             statusInfo = new StatusInfo(UNHEALTHY, "health check failed due to broken node lock");
+        } else if (lastRunStartTimeMillis.get() >= 0 && currentTimeMillisSupplier.getAsLong() -
+            lastRunStartTimeMillis.get() > healthyTimeoutThreshold.millis() + refreshInterval.millis()) {
```

Comment:
@Bukhtawar hm, I think it is still not doing right thing, @getsaurabh02 please correct me if I am wrong. Let me add a few examples:

1. Refresh: every `60s`, Healthy Timeout: `20s`
    In this case the status would not be properly set unless the `flush` takes more than 80s (refresh + healthy)
2. Refresh: every `60s`, Healthy Timeout: `70s`
    In this case the status would not be properly set unless the `flush` takes more than 130s (refresh + healthy)

It is difficult to make any conclusions without flagging somehow that the refresh check is over. The @getsaurabh02 idea https://github.com/opensearch-project/OpenSearch/pull/1167#discussion_r699498772 to reset the start time once the refresh check is over is probably the simplest solution.

What do you think?

Reply:
Yes the problem still exists, as every run in its beginning will also update the ``lastRunStartTimeMillis`` value. As @reta pointed out correctly, the healthy timeout threshold will not be honoured in isolation. 

We should keep states from for run independent as suggested in the previous comment, and reset the value once a run is successful. 

In addition to simplicity, it will also ensure that there are no future bugs, given ``refreshInterval`` is just a node level settings, although it is not dynamic today, may undergo logical changes in future (such as once retries are introduced).

Reply:
Let's say we reset the start time once the refresh check is over, we will still need to wait upto the refresh interval minimum to update the detection start time, since a smaller healthy threshold value would be wrong.

Refresh: every 60s, Healthy Timeout: 20s
In this case we need to wait for 60s as time elapsed since last run will only change on the start of the next run which is the refresh interval. 

Please let me know if I am missing something.
Are you suggesting we have a start and end dedicated variables @reta?

Reply:
> Let's say we reset the start time once the refresh check is over, we will still need to wait upto the refresh interval minimum to update the detection start time, since a smaller healthy threshold value would be wrong.

Ah, yeah, true, with small one we are stuck, correct

> Are you suggesting we have a start and end dedicated variables

That would nail the problem right away, the exact duration of the last flush would be known. The presence of the `start time` will confirm the FS check is started, the presence `stop time` - that FS check is ended (or still ongoing if not set). The only think we have to take care of is synchronization, since start/stop has to be reset simultaneously (otherwise other thread will see incomplete start / stop sequence)

Reply:
Do you think this could be a follow-up for now as synchronization would need more work and additional tests. I'll be happy to work on the improvement ideas @reta?

Reply:
> In this case we need to wait for 60s as time elapsed since last run will only change on the start of the next run which is the refresh interval.

@Bukhtawar What I meant by by resetting the value once a run is successful is inside ``run()`` as below, and not wait for next run to reset. 

```
if (enabled) {
                    setLastRunStartTimeMillis();
                    monitorFSHealth();
                    logger.debug("health check succeeded");
                    // Reset the last run time here as health check completed.
                    resetLastRunStartTimeMillis();
                }
```

Reply:
resetLastRunStartTimeMillis() here updates the ```lastRunStartTimeMillis``` to ```Long.MIN_VALUE``` as suggested in https://github.com/opensearch-project/OpenSearch/pull/1167#discussion_r699498772. Since your evaluation logic in ```getHealth()``` checks for condition ``lastRunStartTimeMillis.get() >= 0 `` at first, you will not evaluate the next condition unless there is an ongoing health check currently.

In case of an ongoing current check, your next condition ``currentTimeMillisSupplier.getAsLong() - lastRunStartTimeMillis.get() > healthyTimeoutThreshold.millis()`` is true if the threshold is breached. 

@Bukhtawar @reta Let me know if this addresses the issue here. Or am I missing something stated earlier.

Reply:
This addresses the problem but makes the logic hard to reason about. I would instead go for an AtomicBoolean to know if there is a check in progress with atomic get and set at the entry of the check and then use the last start time of the run instead. So if only a check is in progress would be consider the last start time of the run and compare against the healthy threshold

Reply:
Yes, only point here is to keep each run independent and have a mechanism to identify an ongoing current check. That could be achieved by multiple ways.

One way as stated above is to have a start timestamp and reset it post completion. Another way as you suggested rightly is by having an Atomic boolean and update it in the beginning and end of each run. Alternatively as @reta suggested with ``start`` and ``end`` atomic time-stamps and asserting ``start > end`` to detect an ongoing run. Either of the them should serve the purpose.

Reply:
@reta @getsaurabh02 Changed are made. Let me know if this works better